tests/test_shadow.py

```diff
@@ -0,0 +1,233 @@
+)
+def test_number_of_samples(
+    observables: list[AbstractBlock], accuracy: float, confidence: float, exp_samples: tuple
+) -> None:
+    N, K = number_of_samples(observables=observables, accuracy=accuracy, confidence=confidence)
+    assert N == exp_samples[0]
+    assert K == exp_samples[1]
+
+
+@pytest.mark.parametrize(
+    "sample, unitary_ids, exp_shadow",
+    [
+        (
+            Counter({"10": 1}),
+            [0, 2],
+            torch.kron(
+                3 * (UNITARY_TENSOR[0].adjoint() @ P1_MATRIX @ UNITARY_TENSOR[0]) - IMAT,
+                3 * (UNITARY_TENSOR[2].adjoint() @ P0_MATRIX @ UNITARY_TENSOR[2]) - IMAT,
+            ),
+        ),
+        (
+            Counter({"0111": 1}),
+            [2, 0, 2, 2],
+            torch.kron(
+                torch.kron(
+                    3 * (UNITARY_TENSOR[2].adjoint() @ P0_MATRIX @ UNITARY_TENSOR[2]) - IMAT,
+                    3 * (UNITARY_TENSOR[0].adjoint() @ P1_MATRIX @ UNITARY_TENSOR[0]) - IMAT,
+                ),
+                torch.kron(
+                    3 * (UNITARY_TENSOR[2].adjoint() @ P1_MATRIX @ UNITARY_TENSOR[2]) - IMAT,
+                    3 * (UNITARY_TENSOR[2].adjoint() @ P1_MATRIX @ UNITARY_TENSOR[2]) - IMAT,
+                ),
+            ),
+        ),
+    ],
+)
+def test_local_shadow(sample: Counter, unitary_ids: list, exp_shadow: Tensor) -> None:
+    shadow = local_shadow(sample=sample, unitary_ids=unitary_ids)
+    assert torch.allclose(shadow, exp_shadow)
+
+
+theta = Parameter("theta")
+
+
+@pytest.mark.skip(reason="Can't fix the seed for deterministic outputs.")
+@pytest.mark.parametrize(
+    "layer, param_values, exp_shadows",
+    [
+        (X(0) @ X(2), {}, [])
+        # (kron(RX(0, theta), X(1)), {"theta": torch.tensor([0.5, 1.0, 1.5])}, [])
+    ],
+)
+def test_classical_shadow(layer: AbstractBlock, param_values: dict, exp_shadows: list) -> None:
+    circuit = QuantumCircuit(2, layer)
+    shadows = classical_shadow(
+        shadow_size=2,
+        circuit=circuit,
+        param_values=param_values,
+    )
+    for shadow, exp_shadow in zip(shadows, exp_shadows):
+        for batch, exp_batch in zip(shadow, exp_shadow):
+            assert torch.allclose(batch, exp_batch, atol=1.0e-2)
+
+
+@pytest.mark.parametrize(
+    "N, K, circuit, param_values, observable, exp_traces",
+    [
+        (2, 1, QuantumCircuit(2, kron(X(0), Z(1))), {}, X(1), torch.tensor([0.0])),
+    ],
+)
+def test_estimators(
+    N: int,
+    K: int,
+    circuit: QuantumCircuit,
+    param_values: dict,
+    observable: AbstractBlock,
+    exp_traces: Tensor,
+) -> None:
+    shadows = classical_shadow(shadow_size=N, circuit=circuit, param_values=param_values)
+    estimated_traces = estimators(
+        qubit_support=circuit.block.qubit_support,
+        N=N,
+        K=K,
+        shadow=shadows[0],
+        observable=observable,
+    )
+    assert torch.allclose(estimated_traces, exp_traces)
+
+
+# @pytest.mark.flaky(max_runs=5)
```

Comment:
Isn't it possible to enable these tests ?

Reply:
Yes I forgot to uncomment when testing as these are long.